### ISSUE TITLE
Using npm modules in web workers in our code - a retrospective

### ISSUE THREAD

**Issue by TylerLeonhardt** [microsoft/vscode#130302]
This is kind of a meta issue around my experience adding a web worker to our code for automatic language detection. If you don't know what this feature is all about here's a gif of it in action:
![thing](https://user-images.githubusercontent.com/2644648/128579684-29c21315-6448-4304-a491-320307133bfc.gif)
It uses a ML model to guess what language your code is and sets the mode accordingly automatically. Because running an ML model can be intensive, we decided to have it run in a web worker. It's the perfect place for a feature like this.

The model is brought in using a npm module found here (and published on npm):  https://github.com/Microsoft/vscode-languagedetection
and before moving to a worker, we were able to simply use this in the service's code and everything worked as expected:
```ts
await import('@vscode/vscode-languagedetection');
```
> [code pointer in history](https://github.com/microsoft/vscode/blob/3d9899db8ea9c7b70dcf8379a952c8d554d607b9/src/vs/workbench/services/languageDetection/browser/languageDetectionService.ts#L64)

However, moving this code over to the web worker did not have the same success... with a very simple error of:
```
TypeError: Failed to fetch
```
The network tab gives us a hint:
![image](https://user-images.githubusercontent.com/2644648/128580432-17a11f76-7942-4a3a-85c2-21b71b8018ba.png)
and if I hover over the file, the path is:
```
vscode-file://vscode-app/Users/tyleonha/Code/Microsoft/vscode/out/@vscode/vscode-languagedetection.js
```
This path is actually wrong. It should have come from a `node_modules` folder... but it does not. What's also interesting is that this issue only shows up in Electron... in the web it works just fine:
![image](https://user-images.githubusercontent.com/2644648/128580332-e890404f-445a-4926-a1aa-f9cb0ff027f1.png)

Now let's talk about the workaround... in order to get this working I used the same tactic that I use for loading the non-JS files of the model (look at the additional context below). By using the `FileAccess` API and a relative path, you can always get the Uri to any file in our package, and that's what I do to get the JS file that we import:

Do this on the UI side:
```ts
const moduleLocation = '../../../../../../node_modules/@vscode/vscode-languagedetection';
const moduleLocationAsar = '../../../../../../node_modules.asar/@vscode/vscode-languagedetection';
this._environmentService.isBuilt && !isWeb
    ? FileAccess.asBrowserUri(`${moduleLocationAsar}/dist/lib/index.js`, require).toString(true)
    : FileAccess.asBrowserUri(`${moduleLocation}/dist/lib/index.js`, require).toString(true),
```
And then use a fmr (basically a request to the UI side) on the worker side to get the Uri and import it:
```ts
const uri: string = await this._host.fhr('getIndexJsUri', []);
const { ModelOperations } = await import(uri) as typeof import('@vscode/vscode-languagedetection');
```

This........ works...... but it's really quite ugly and feel like there's gotta be a better way to do this. I think the folks that can help the most is @alexdima and @deepak1556 ... but maybe @bpasero and @rebornix are also interested in this knowledge.

The main PR that moved the language detection to a worker is found here (there are few additional misc commits that have gone in since then): #130006 

### Additional context

The logic around firing the model is also a bit odd...
1. the code can be found here: https://github.com/Microsoft/vscode-languagedetection this repo actually contains the model
2. Since the model is not purely written in JS (it consists of 2 files: a [JSON](https://github.com/microsoft/vscode-languagedetection/blob/main/model/model.json) file & a [binary](https://github.com/microsoft/vscode-languagedetection/blob/main/model/group1-shard1of1.bin) file) it's very hard to bundle these files properly such that it works in Node.js, Electron, & web. Because of this, we have to load them using VS Code's [FileAccess API](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts#L47-L51) (Similar tactics are used to load in vscode-oniguruma's WASM file)

**Comment by deepak1556** [comment]
The issue is not impacted by the protocol used for the resource its just a limitation from today's loader, `vscode-loader` supports [different loaders](https://github.com/microsoft/vscode-loader/blob/cbe8f639e7e2ae8a0ede5da411555815158bd13e/src/core/scriptLoader.ts#L53-L80) depending on the environment in which a script resource is included, the current `WorkerScriptLoader` does not respect node modules as it [cannot use](https://github.com/microsoft/vscode-loader/blob/main/src/core/env.ts#L104-L106) the `NodeScriptLoader` inside a web worker, so it ends up resolving similar to other workbench resources and passes it down to the [fetch](https://github.com/microsoft/vscode-loader/blob/cbe8f639e7e2ae8a0ede5da411555815158bd13e/src/core/scriptLoader.ts#L177) api.


One alternative workaround for your scenario that will work today is to specify a `baseUrl` inside `languageDetectionSimpleWorker.ts` that will be [used by the loader](https://github.com/microsoft/vscode-loader/blob/cbe8f639e7e2ae8a0ede5da411555815158bd13e/src/core/configuration.ts#L481-L483) in resolving the resource.

```typescript
// On the UI side
const moduleUri = fileUriFromPath(environmentService.appRoot, { scheme: 'vscode-file', fallbackAuthority: 'vscode-app' })/node_modules;

// On the worker
const uri: string = await this._host.fhr('getModuleUri', []);
let loaderConfiguration: any = (<any>self).require.getConfig();
loaderConfiguration.baseUrl = uri;
(<any>self).require.config(loaderConfiguration);
const { ModelOperations } = await import('@vscode/vscode-languagedetection/dist/lib/index');
```

I am not sure what is the path forward for resources loaded from `node_modules.asar` in the workbench as we move close to the browser model in which case a solution would be needed there as well. @alexdima @bpasero thoughts ?

**Comment by alexdima** [comment]
I agree this is ugly, but it stems from the way [AMD](https://github.com/amdjs/amdjs-api/wiki/AMD) works. In AMD, when someone wants to load a module (e.g. `require(['vs/base/browser/dom'], ...)`), the AMD loader will first need to convert this module identifier into a path. 

The transformation of a module identifier to a path can be tweaked via loader configurations. The loader uses two main configurations, `baseUrl` and `paths` for this purpose, but there are others. Given the loader config:
```
{ 
  baseUrl: 'vscode-file://vscode-app/Users/tyleonha/Code/Microsoft/vscode/out/', 
  paths: {} 
}
```
The loader will transform the module id `vs/base/browser/dom` to:
```
vscode-file://vscode-app/Users/tyleonha/Code/Microsoft/vscode/out/vs/base/browser/dom.js
```
But you can see how this falls apart for node modules. As you have documented above, the module id `@vscode/vscode-languagedetection` will similarly be transformed to the path:
```
vscode-file://vscode-app/Users/tyleonha/Code/Microsoft/vscode/out/@vscode/vscode-languagedetection.js
```
which does not exist.

So the error you are seeing makes perfect sense in a strict AMD compliant loader, which is how `vscode-loader` operates inside a web worker.

---

There are two distinct reasons why `require(['@vscode/vscode-languagedetection'])` works on the desktop in the renderer process and works on the web in the main page, but not on the worker.

---

On the desktop, in the renderer process, `vscode-loader` is not working as an AMD spec-compliant loader. When running with nodejs integration (which is how we normally run it), the loader is configured with the setting `amdModulesPattern: /^vs\//`. This setting tells `vscode-loader` that only module ids starting with `vs/` are AMD modules and the rest of the module ids refer to node modules. So in the desktop, in the non-sandboxed case, the loader will delegate all other `require` calls to the underlying nodejs commonjs based `require`.

You can see this configuration [here](https://github.com/microsoft/vscode/blob/00de9bb8aee3704a81b370d12704f4582d4e517b/src/bootstrap-window.js#L131-L144). Since maybe you are there, you can see how things work when there is no nodejs integration (e.g. for the sandbox case or for the web case).

The web is configured [here](https://github.com/microsoft/vscode/blob/00de9bb8aee3704a81b370d12704f4582d4e517b/src/vs/code/browser/workbench/workbench-dev.html#L48). The global `self.webPackagePaths` comes from `out/vs/webPackagePaths.js`, which gets generated by running `yarn watch`. Here is how it looks like:
```
self.webPackagePaths = {
  "@microsoft/applicationinsights-web": "dist/applicationinsights-web.js",
  "@vscode/vscode-languagedetection": "dist/lib/index.js",
  "iconv-lite-umd": "lib/iconv-lite-umd.js",
  "jschardet": "dist/jschardet.min.js",
  "tas-client-umd": "lib/tas-client-umd.js",
  "vscode-oniguruma": "release/main.js",
  "vscode-textmate": "release/main.js",
  "xterm": "lib/xterm.js",
  "xterm-addon-search": "lib/xterm-addon-search.js",
  "xterm-addon-unicode11": "lib/xterm-addon-unicode11.js",
  "xterm-addon-webgl": "lib/xterm-addon-webgl.js"
};
```

Basically, both on the desktop with sandbox case and in the web case, the loader, since it doesn't have access to nodejs is configured using `paths` to how to transform a module id to a path. 

> Note: now that I see the sandbox config paths, it looks like they are out of sync. cc @lramos15 we could also investigate using the generated `webPackagePaths` in `bootstrap-window` for the sandboxed case.

---

So long story short, if you want to load a node module in a web worker (you're the first one that wants to do that, sorry!), you need to configure the loader and teach it what is the "main" file of that node module. Perhaps something like:
```
require.config({
  paths: {
    '@vscode/vscode-languagedetection': '../node_modules/@vscode/vscode-languagedetection/dist/lib/index.js'
  }
});
```

**Comment by lramos15** [comment]
@alexdima Does bootstrap-window use the same packages as web?

**Comment by alexdima** [comment]
> Does bootstrap-window use the same packages as web?

Short answer is I don't know.

In a sandboxed renderer, we load the services implemented in `/electron-sandbox/` or in `/browser/`, so there might be differences in what node modules those different implementations end up loading. I think there is a way to launch (maybe `--sanbox` if I'm not mistaken) to test this. @deepak1556 @bpasero what is the flag to launch with the sandbox enabled?

**Comment by bpasero** [comment]
Yes, we load the same node modules in web as we do in desktop, there is no difference anymore. Especially since we moved to `vscode-file` there is also no node.js involved anymore in loading modules.

**Comment by bpasero** [comment]
Hm, that said I wonder if this code can actually go away given we use `script` tags now:

https://github.com/microsoft/vscode/blob/b6ca3e52ae62b64f7e30009bed4402a0235a4e92/src/bootstrap-window.js#L132-L141

**Comment by alexdima** [comment]
> Hm, that said I wonder if this code can actually go away given we use script tags now:

@bpasero I think the short answer is no. But maybe it can be replaced by using the generated `out/vs/webPackagePaths.js`, but that might be difficult to bundle. If a loader is not configured with paths for node modules, it will go looking for them using the AMD lookup rules. The long answer is [here](https://github.com/microsoft/vscode/issues/130302#issuecomment-910383180).

**Comment by bpasero** [comment]
@alexdima oh wow, so this is a bit of a missing adoption then. We indeed load all node modules in the renderer still via node.js `require` and thus we do not benefit from the V8 caching support here that we have for the other code.

I just gave it a quick test and changed the code to:

```js
loaderConfig.paths = {
	'vscode-textmate': `../node_modules/vscode-textmate/release/main`,
	'vscode-oniguruma': `../node_modules/vscode-oniguruma/release/main`,
	'xterm': `../node_modules/xterm/lib/xterm.js`,
	'xterm-addon-search': `../node_modules/xterm-addon-search/lib/xterm-addon-search.js`,
	'xterm-addon-unicode11': `../node_modules/xterm-addon-unicode11/lib/xterm-addon-unicode11.js`,
	'xterm-addon-webgl': `../node_modules/xterm-addon-webgl/lib/xterm-addon-webgl.js`,
	'iconv-lite-umd': `../node_modules/iconv-lite-umd/lib/iconv-lite-umd.js`,
	'jschardet': `../node_modules/jschardet/dist/jschardet.min.js`,
};

if (!safeProcess.sandboxed) {
	loaderConfig.amdModulesPattern = /^vs\//;
}
```

But that would still result in node modules getting loaded via `require`, probably because the `amdModulesPattern` rule is winning over the configured paths. I cannot remove the `amdModulesPattern` rule though because we still have a few places that require node.js standard modules (like `fs`).

Is it worth doing something now or do this after we have gotten rid of the few node.js modules we still load (mainly file service which I wanted to look at this iteration).

**Comment by alexdima** [comment]
:+1: If we want that node modules get loaded as AMD, then we need to either remove `amdModulesPattern` or adjust it to a regex that contains all AMD node modules. If we remove `amdModulesPattern`, there will be a slight performance loss or a 404, as requiring `fs` would first look for it in `out/fs.js`. We can remove `amdModulesPattern` and use a config called `nodeModules` to list one by one the modules we want to be delegated to nodejs. Complete example:
```
loaderConfig.paths = {
	'vscode-textmate': `../node_modules/vscode-textmate/release/main`,
	'vscode-oniguruma': `../node_modules/vscode-oniguruma/release/main`,
	'xterm': `../node_modules/xterm/lib/xterm.js`,
	'xterm-addon-search': `../node_modules/xterm-addon-search/lib/xterm-addon-search.js`,
	'xterm-addon-unicode11': `../node_modules/xterm-addon-unicode11/lib/xterm-addon-unicode11.js`,
	'xterm-addon-webgl': `../node_modules/xterm-addon-webgl/lib/xterm-addon-webgl.js`,
	'iconv-lite-umd': `../node_modules/iconv-lite-umd/lib/iconv-lite-umd.js`,
	'jschardet': `../node_modules/jschardet/dist/jschardet.min.js`,
};

loaderConfig.nodeModule = ['fs', 'os', 'child_process', 'net', ...];
```

**Comment by bpasero** [comment]
@alexdima thanks, I went with the second approach:

https://github.com/microsoft/vscode/blob/ed5c561ff33cf0f29cc06726669225a338d4272d/src/bootstrap-window.js#L127-L145

**Comment by TylerLeonhardt** [comment]
@bpasero  with that change should I switch over to requiring vscode-languagedetection using its name now?

**Comment by bpasero** [comment]
@TylerLeonhardt not sure I can follow, I tried to copy exact the name of our `import`: `@vscode/vscode-languagedetection`, e.g. 

https://github.com/microsoft/vscode/blob/430e98bf678e9b0e8f69365c87688dfd3ceb9778/src/vs/workbench/services/languageDetection/browser/languageDetectionSimpleWorker.ts#L6-L6

or

https://github.com/microsoft/vscode/blob/430e98bf678e9b0e8f69365c87688dfd3ceb9778/src/vs/workbench/services/languageDetection/browser/languageDetectionSimpleWorker.ts#L52-L52

**Comment by TylerLeonhardt** [comment]
The second one is the workaround for this issue. The first is just an `import type` so that the types in the file aren't a mess.

**Comment by bpasero** [comment]
Ok I am not sure I follow, but if you want to change the name of your module, wouldn't it be better to rename it in NPM?

**Comment by TylerLeonhardt** [comment]
In the original description of the issue, I talk about how I can't simply `await import('@vscode/vscode-languagedetection')` because the webworker thinks the location of that package is somewhere else.

So the workaround was to pass the uri to the index.js to the webworker so the webworker could load the module correctly.

I want to get rid of that hacky "passing the path" logic and replace it with:

`await import('@vscode/vscode-languagedetection')`

And was wondering if your commit fixed that

**Comment by bpasero** [comment]
Short: Oh I see, sorry my change is totally unrelated. I just hijacked this issue to clarify something I had forgotten to follow up with.

Long: we are in the process of building a node.js free renderer and part of it is to load all code via `script` tags and not node.js `require` calls. Because I had forgotten to tell the loader where the node modules are on disk, the loader was falling back to using `require` from node.js to load them. With my change we should now see `script` tags for node modules as well and the loader will no longer use node.js to load these modules.

**Comment by alexdima** [comment]
> I want to get rid of that hacky "passing the path" logic and replace it with:

@TylerLeonhardt The loader configuration is [forwarded](https://github.com/microsoft/vscode/blob/c99edb1545caba6c970ea5c861cb23672bcf3882/src/vs/base/common/worker/simpleWorker.ts#L233-L241) from the renderer process to the web worker.

So if @bpasero pushes some code which changes the loader config to contain a path configuration for `@vscode/vscode-languagedetection`, then that configuration will be transported to the web worker. 

From https://github.com/microsoft/vscode/issues/130302#issuecomment-910383180:
```
require.config({
  paths: {
    '@vscode/vscode-languagedetection': '../node_modules/@vscode/vscode-languagedetection/dist/lib/index.js'
  }
});
```

**Comment by bpasero** [comment]
@alexdima in today's insider I notice a regression from my change:

![image](https://user-images.githubusercontent.com/900690/131990038-603b0e85-f5cb-4275-bc83-f1a8e9cfe01a.png)

All modules fail to fetch, e.g. in more detail:

![image](https://user-images.githubusercontent.com/900690/131990084-2d9ef7fc-9a0e-47f9-bac0-342aa97cb7c7.png)

I wonder if this is a ASAR problem or a general problem with using the script tags for anything that is inside asar?

**Comment by deepak1556** [comment]
Files from asar will be respected by the browser network code path, the issue above is the path being incorrect. It should be `node_modules.asar` instead of `node_modules`, then only the network code will use the asar resource loader in the runtime.

**Comment by alexdima** [comment]
oh no :).

Yes, the loader running in a nodejs enabled Electron renderer will first try to load a module using the AMD rules (in this case respecting the paths setting you have defined), and if that fails, then it also tries loading the module using nodejs. So we are super lucky.

I also assume the problem is that the files are in an ASAR.

**Comment by deepak1556** [comment]
For ex:

Should be 

```
vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/node_modules.asar/tas-client-umd/lib/tas-client-umd.js
```

instead of 

```
vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/node_modules/tas-client-umd/lib/tas-client-umd.js
```

**Comment by alexdima** [comment]
To further explain, the nodejs require would also not be able to find them and would throw an error. That is why we patch `Module._resolveLookupPaths` [here](https://github.com/microsoft/vscode/blob/70e30bc64e9ec7dd996259a00765f62a5e339f59/src/bootstrap.js#L86). We actually inject `node_modules.asar` as a lookup path.

So to get script loading working, we would ned to massage the configured `paths` to be `node_modules.asar`, as @deepak1556  mentions.

**Comment by bpasero** [comment]
Thanks! I changed it to this now:

https://github.com/microsoft/vscode/blob/f8eb737b8286b8062036606ab760f4a59e58b333/src/bootstrap-window.js#L130-L141

**Comment by lramos15** [comment]
@bpasero Would be interesting if you could use this function from the build script

https://github.com/microsoft/vscode/blob/d2a8f91fb5107d4bf677083a84546ffed1957f4d/build/lib/util.ts#L344-L367
since that's what we use to populate all the workbench.html's and the loader. You would then just be doing something like 
```js
		Object.keys(webPackagePaths).map(function (key, index) {
			webPackagePaths[key] = `../node_modules.asar/${key}/${webPackagePaths[key]}`;
		});
```
and benefit from the auto detection of new modules

**Comment by bpasero** [comment]
@lramos15 thats cool but we cannot use something like `fs.readFileSync` in the node.js free renderer going forward.

**Comment by lramos15** [comment]
> @lramos15 thats cool but we cannot use something like `fs.readFileSync` in the node.js free renderer going forward.

It is built and bundled with vscode, but I'm not sure the process needed to access those file contents without being dependent on node.

**Comment by TylerLeonhardt** [comment]
It's been a while since I looked at this, and there's a lot going on here so I'm just going to close it. Something that I think might be useful to understand in the future is how could we share node modules tensorflow.js being one of them. Right now `@vscode/vscode-languagedetection` _bundles it in_ but it's big... and maybe some _other_ ML experience that also uses tensorflow will get added to VS Code and ideally we'd want to share to make the bundle smaller.
